apps/labs/posts/scaling-asyncio-on-free-threaded-python.md

```diff
@@ -0,0 +1,169 @@
+   currently running, and once the task completes or suspends by awaiting on
+   something, the current task state is reset. High-level APIs such as
+   `asyncio.timeout()` and `asyncio.TaskGroup` rely on current task for proper
+   cancellation of tasks. The current task can be accessed using
+   `asyncio.current_task()`.
+
+Up until now, `asyncio` was designed with the assumption of a single-threaded
+environment, and relied on the GIL to manage access to shared state. The current
+task state was stored in a global dictionary mapping threads to their current
+task, and all tasks were stored in a global `WeakSet`. This scales poorly with
+the number of threads in free-threading because of reference counting and
+locking contention on these global data structures.
+
+In Python 3.14, I have implemented several changes to fix thread safety of
+`asyncio` and enable it to scale effectively on the free-threaded build of
+CPython. It is now implemented using lock-free data structures and per-thread
+state, allowing for highly efficient task management and execution across
+multiple threads. In the general case of multiple event loops running in
+parallel, there is no lock contention and scales linearly with the number of
+threads.
+
+Here are the key changes:
+
+1. **Per-thread linked list of tasks**:
+   Python 3.14 introduces a per-thread circular double linked list implementation
+   for storing tasks instead of global `WeakSet`. The linked list is per-thread,
+   meaning that each thread maintains its own list of tasks and allows for
+   lock-free addition and removal of tasks. The use of weak references is
+   removed entirely which was slow and prone to contention, by instead making
+   tasks responsible for removing themselves from the list when they are done.
+   This requires cooperation between task's deallocator and the executing
+   threads to ensure that the task is removed from the list before it is freed,
+   otherwise a thread could try accessing already freed task. By removing the
+   use of weak references, the overhead of reference counting is eliminated
+   entirely and addition/removal of task in the list now requires only updating
+   the pointers in the linked list.
+
+   This design allows for efficient, lock-free and thread-safe task management
+   and scales well with the number of threads in free-threading.
+
+2. **Per-thread current task**:
+   Python 3.14 stores the current task on the current thread state instead of a
+   global dictionary mapping event loops to their current tasks. By storing the
+   current task on thread state, the overhead of accessing the current task is
+   reduced, and it allows for lock-free access to the current task while
+   avoiding dictionary lookup. It allows for faster switching between tasks
+   which is a very frequent operation in asyncio.
+
+   This design allows for lock-free access to the current task and avoids
+   reference counting and lock contention on the global dictionary.
```

Comment:
I'm not sure how easy it is to collect the PRs that went into each of these two items, but maybe if it's a reasonable task you could include links to the PRs that implemented each of these?

Reply:
Added links to the PRs.